contracts/src/Assets.sol

```diff
@@ -159,6 +194,88 @@ library Assets {
         emit IGateway.TokenSent(token, sender, destinationChain, destinationAddress, amount);
     }
 
+    function _sendForeignTokenCosts(
+        ParaID destinationChain,
+        uint128 destinationChainFee,
+        uint128 maxDestinationChainFee
+    ) internal view returns (Costs memory costs) {
+        AssetsStorage.Layout storage $ = AssetsStorage.layout();
+        if ($.assetHubParaID == destinationChain) {
+            costs.foreign = $.assetHubReserveTransferFee;
+        } else {
+            // Reduce the ability for users to perform arbitrage by exploiting a
+            // favourable exchange rate. For example supplying Ether
+            // and gaining a more valuable amount of DOT on the destination chain.
+            //
+            // Also prevents users from mistakenly sending more fees than would be required
+            // which has negative effects like draining AssetHub's sovereign account.
+            //
+            // For safety, `maxDestinationChainFee` should be less valuable
+            // than the gas cost to send tokens.
+            if (destinationChainFee > maxDestinationChainFee) {
+                revert InvalidDestinationFee();
+            }
+
+            // If the final destination chain is not AssetHub, then the fee needs to additionally
+            // include the cost of executing an XCM on the final destination parachain.
+            costs.foreign = $.assetHubReserveTransferFee + destinationChainFee;
+        }
+        // We don't charge any extra fees beyond delivery costs
+        costs.native = 0;
+    }
+
+    // @dev Transfer Polkadot-native tokens back to Polkadot
+    function _sendForeignToken(
+        bytes32 foreignID,
+        address token,
+        address sender,
+        ParaID destinationChain,
+        MultiAddress calldata destinationAddress,
+        uint128 destinationChainFee,
+        uint128 maxDestinationChainFee,
+        uint128 amount
+    ) internal returns (Ticket memory ticket) {
+        AssetsStorage.Layout storage $ = AssetsStorage.layout();
+
+        Token(token).burn(sender, amount);
+
+        ticket.dest = $.assetHubParaID;
+        ticket.costs = _sendForeignTokenCosts(destinationChain, destinationChainFee, maxDestinationChainFee);
+
+        // Construct a message payload
+        if (destinationChain == $.assetHubParaID) {
+            // The funds will be minted into the receiver's account on AssetHub
+            if (destinationAddress.isAddress32()) {
+                // The receiver has a 32-byte account ID
+                ticket.payload = SubstrateTypes.SendTokenToAssetHubAddress32(
+                    token, destinationAddress.asAddress32(), $.assetHubReserveTransferFee, amount
+                );
```

Comment:
I would always expect `foreignID` for PNA, not the token address.

https://github.com/Snowfork/polkadot-sdk/blob/f1155b124813eb535e19e0ad5d2b31f0135b6a8b/bridges/snowbridge/primitives/router/src/inbound/mod.rs#L63

Reply:
Fixed in https://github.com/Snowfork/snowbridge/pull/1275/commits/57fa08c1ad101f46a12fc469b75d30bd76658290